ampd/src/handlers/multisig.rs

```diff
@@ -52,23 +60,126 @@ where
         .collect()
 }
 
+pub struct Handler<B>
+where
+    B: BroadcasterClient,
+{
+    worker: TMAddress,
+    multisig: TMAddress,
+    broadcaster: B,
+    signer: SharableEcdsaClient,
+}
+
+impl<B> Handler<B>
+where
+    B: BroadcasterClient,
+{
+    pub fn new(
+        worker: TMAddress,
+        multisig: TMAddress,
+        broadcaster: B,
+        signer: SharableEcdsaClient,
+    ) -> Self {
+        Self {
+            worker,
+            multisig,
+            broadcaster,
+            signer,
+        }
+    }
+
+    async fn broadcast_signature(
+        &self,
+        session_id: Uint64,
+        signature: HexBinary,
+    ) -> error_stack::Result<(), Error> {
+        let msg = serde_json::to_vec(&ExecuteMsg::SubmitSignature {
+            session_id,
+            signature,
+        })
+        .expect("submit signature msg should serialize");
+
+        let tx = MsgExecuteContract {
+            sender: self.worker.as_ref().clone(),
+            contract: self.multisig.as_ref().clone(),
+            msg,
+            funds: vec![],
+        };
+
+        self.broadcaster
+            .broadcast(tx)
+            .await
+            .change_context(Error::Broadcaster)
+    }
+}
+
+#[async_trait]
+impl<B> EventHandler for Handler<B>
+where
+    B: BroadcasterClient + Send + Sync,
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &events::Event) -> error_stack::Result<(), Error> {
+        let SigningStartedEvent {
+            contract_address,
+            session_id,
+            pub_keys,
+            msg,
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(());
+            }
+            result => result.change_context(DeserializeEvent)?,
+        };
+
+        if self.multisig != contract_address {
+            return Ok(());
+        }
+
```

Comment:
Can we add a logline here? Just something indicating a signing started event was received. Ampd was quite silent last time I ran it, and it was pretty hard to figure out if it was working or not